[input.smart]: Specifying multiple drives in the devices= parameter fails when the '-d' parameter is used to differentiate drives within a hardware RAID array

### Relevant telegraf.conf:
<!-- Place config in the toml code section. -->
```toml
[[inputs.smart]]
path_smartctl = "/usr/sbin/smartctl"
path_nvme = "/usr/sbin/nvme"
enable_extensions = ["auto-on"]
use_sudo = true
attributes = true
#
#  List all 24 drives within the Areca RAID controller:
devices = [ "dev/sg3 -d areca,2/2", "/dev/sg3 -d areca,3/2"]
#
#  Note that the above was concatenated for brevity.  The remainder of that line is below for all drives:
#, "/dev/sg3 -d areca,3/2", "/dev/sg3 -d areca,4/2", "/dev/sg3 -d areca,5/2", "/dev/sg3 -d areca,6/2", "/dev/sg3 -d areca,7/2", "/dev/sg3 -d areca,8/2", "/dev/sg3 -d areca,9/2", "/dev/sg3 -d areca,10/2", "/dev/sg3 -d areca,11/2", "/dev/sg3 -d areca,12/2", "/dev/sg3 -d areca,13/2", "/dev/sg3 -d areca,14/2", "/dev/sg3 -d areca,15/2", "/dev/sg3 ->
```

### System info:

Telegraf 1.17.0 (git: HEAD 3f7a54c9)
Linux sophie 5.4.0-58-generic #64-Ubuntu SMP (Ubuntu 20.04)
smartmontools release 7.1 dated 2019-12-30 at 15:00:11 UTC
smartmontools SVN rev 5022 dated 2019-12-30 at 15:00:49
smartmontools build host: x86_64-pc-linux-gnu

### Docker
Docker not being used.

### Steps to reproduce:

<!-- Describe the steps to reproduce the bug. -->

1. Trying to specify the individual drives within the Areca RAID array from within the [input.smart] section of telegraf.conf fails
2. Any one of the following lines works:
```
devices = [ "/dev/sg3 -d areca,2/2" ]             # A single drive from the Areca array, indicated via 'areca,2/2'
devices = [ "/dev/sg3 -d areca,7/2" ]           # A different drive from the Areca array - 'areca 7/2'
devices = [ "/dev/sg3 -d areca,1/2", "/dev/sdd" ]   # Multiple drives, so long as only one is from the Areca array
devices = [ "/dev/sdd", "/dev/sdd"]                # Duplicating the same non-Areca drive seems to work without error

```
3.  But the following doesn't work:
`devices = [ "/dev/sg3 -d areca,2/2", "/dev/sg3 -d areca,3/2" ]   # Two drives from within the Areca array`

### Expected behaviour:

Smartctl can retrieve data from individual drives within a hardware array provided the '--device=' parameter is correct.
For Areca arrays, the following is correct syntax:

`smartctl --info --health --attributes --tolerance=verypermissive -n standby --format=brief --device=areca,1/2 /dev/sg3`

This returns comprehensive data about the specific drive within the array:

```
root@sophie:/etc/telegraf# smartctl --info --health --attributes --tolerance=verypermissive -n standby --format=brief --device=areca,1/2 /dev/sg3
smartctl 7.1 2019-12-30 r5022 [x86_64-linux-5.4.0-58-generic] (local build)
Copyright (C) 2002-19, Bruce Allen, Christian Franke, www.smartmontools.org

=== START OF INFORMATION SECTION ===
Device Model:     HGST HDN721010ALE604
Serial Number:    1SJS3J5Z
LU WWN Device Id: 5 000cca 26be6b0c3
Firmware Version: 83XN
User Capacity:    10,000,831,348,736 bytes [10.0 TB]
Sector Sizes:     512 bytes logical, 4096 bytes physical
Rotation Rate:    7200 rpm
Form Factor:      3.5 inches
Device is:        Not in smartctl database [for details use: -P showall]
ATA Version is:   ACS-2, ATA8-ACS T13/1699-D revision 4
SATA Version is:  SATA 3.2, 6.0 Gb/s (current: 6.0 Gb/s)
Local Time is:    Wed Jan 13 01:13:50 2021 PST
SMART support is: Available - device has SMART capability.
SMART support is: Enabled
Power mode is:    ACTIVE or IDLE

=== START OF READ SMART DATA SECTION ===
SMART overall-health self-assessment test result: PASSED

SMART Attributes Data Structure revision number: 16
Vendor Specific SMART Attributes with Thresholds:
ID# ATTRIBUTE_NAME          FLAGS    VALUE WORST THRESH FAIL RAW_VALUE
  1 Raw_Read_Error_Rate     PO-R--   100   100   016    -    0
  2 Throughput_Performance  --S---   135   135   054    -    92
  3 Spin_Up_Time            POS---   175   175   024    -    406 (Average 344)
  4 Start_Stop_Count        -O--C-   100   100   000    -    99
  5 Reallocated_Sector_Ct   PO--CK   100   100   005    -    0
  7 Seek_Error_Rate         -O-R--   100   100   067    -    0
  8 Seek_Time_Performance   --S---   128   128   020    -    18
  9 Power_On_Hours          -O--C-   098   098   000    -    16232
 10 Spin_Retry_Count        -O--C-   100   100   060    -    0
 12 Power_Cycle_Count       -O--CK   100   100   000    -    99
 22 Unknown_Attribute       PO---K   100   100   025    -    100
192 Power-Off_Retract_Count -O--CK   100   100   000    -    823
193 Load_Cycle_Count        -O--C-   100   100   000    -    823
194 Temperature_Celsius     -O----   166   166   000    -    36 (Min/Max 17/59)
196 Reallocated_Event_Count -O--CK   100   100   000    -    0
197 Current_Pending_Sector  -O---K   100   100   000    -    0
198 Offline_Uncorrectable   ---R--   100   100   000    -    0
199 UDMA_CRC_Error_Count    -O-R--   200   200   000    -    0
```

When a different value is used, Smartctl correctly retrieves the data for the different disk within the array:

```
root@sophie:/etc/telegraf# smartctl --info --health --attributes --tolerance=verypermissive -n standby --format=brief --device=areca,7/2 /dev/sg3
smartctl 7.1 2019-12-30 r5022 [x86_64-linux-5.4.0-58-generic] (local build)
Copyright (C) 2002-19, Bruce Allen, Christian Franke, www.smartmontools.org

=== START OF INFORMATION SECTION ===
Device Model:     HGST HDN721010ALE604
Serial Number:    1SJTUGKZ
LU WWN Device Id: 5 000cca 26be77786
Firmware Version: 83XN
User Capacity:    10,000,831,348,736 bytes [10.0 TB]
Sector Sizes:     512 bytes logical, 4096 bytes physical
Rotation Rate:    7200 rpm
Form Factor:      3.5 inches
Device is:        Not in smartctl database [for details use: -P showall]
ATA Version is:   ACS-2, ATA8-ACS T13/1699-D revision 4
SATA Version is:  SATA 3.2, 6.0 Gb/s (current: 6.0 Gb/s)
Local Time is:    Wed Jan 13 01:14:54 2021 PST
SMART support is: Available - device has SMART capability.
SMART support is: Enabled
Power mode is:    ACTIVE or IDLE

=== START OF READ SMART DATA SECTION ===
SMART overall-health self-assessment test result: PASSED

SMART Attributes Data Structure revision number: 16
Vendor Specific SMART Attributes with Thresholds:
ID# ATTRIBUTE_NAME          FLAGS    VALUE WORST THRESH FAIL RAW_VALUE
  1 Raw_Read_Error_Rate     PO-R--   100   100   016    -    0
  2 Throughput_Performance  --S---   134   134   054    -    96
  3 Spin_Up_Time            POS---   178   178   024    -    316 (Average 424)
  4 Start_Stop_Count        -O--C-   100   100   000    -    109
  5 Reallocated_Sector_Ct   PO--CK   100   100   005    -    0
  7 Seek_Error_Rate         -O-R--   100   100   067    -    0
  8 Seek_Time_Performance   --S---   128   128   020    -    18
  9 Power_On_Hours          -O--C-   098   098   000    -    20286
 10 Spin_Retry_Count        -O--C-   100   100   060    -    0
 12 Power_Cycle_Count       -O--CK   100   100   000    -    109
 22 Unknown_Attribute       PO---K   100   100   025    -    100
192 Power-Off_Retract_Count -O--CK   100   100   000    -    1020
193 Load_Cycle_Count        -O--C-   100   100   000    -    1020
194 Temperature_Celsius     -O----   166   166   000    -    36 (Min/Max 16/62)
196 Reallocated_Event_Count -O--CK   100   100   000    -    0
197 Current_Pending_Sector  -O---K   100   100   000    -    0
198 Offline_Uncorrectable   ---R--   100   100   000    -    0
199 UDMA_CRC_Error_Count    -O-R--   200   200   000    -    0
```

The [input.smart] plugin allows for specifying device and device type:
```
#   ## Optionally specify devices and device type, if unset
#   ## a scan (smartctl --scan and smartctl --scan -d nvme) for S.M.A.R.T. devices will be done
#   ## and all found will be included except for the excluded in excludes.
#   # devices = [ "/dev/ada0 -d atacam", "/dev/nvme0"]
devices = [ "/dev/sg3 -d areca,12/2" ]

```
The above works correctly:
```
root@sophie:/etc/telegraf# nano telegraf.conf
root@sophie:/etc/telegraf# telegraf  --test| grep smart
2021-01-13T09:17:43Z I! Starting Telegraf 1.17.0
2021-01-13T09:17:43Z I! Using config file: /etc/telegraf/telegraf.conf
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=POSR--,host=sophie,id=1,model=ST10000NM0086-2AA101,name=Raw_Read_Error_Rate,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=344184i,threshold=44i,value=100i,worst=64i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=PO----,host=sophie,id=3,model=ST10000NM0086-2AA101,name=Spin_Up_Time,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=0i,threshold=0i,value=93i,worst=92i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--CK,host=sophie,id=4,model=ST10000NM0086-2AA101,name=Start_Stop_Count,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=51i,threshold=20i,value=100i,worst=100i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=PO--CK,host=sophie,id=5,model=ST10000NM0086-2AA101,name=Reallocated_Sector_Ct,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=48i,threshold=10i,value=100i,worst=100i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=POSR--,host=sophie,id=7,model=ST10000NM0086-2AA101,name=Seek_Error_Rate,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=427333388i,threshold=45i,value=86i,worst=60i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--CK,host=sophie,id=9,model=ST10000NM0086-2AA101,name=Power_On_Hours,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=13568i,threshold=0i,value=85i,worst=85i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=PO--C-,host=sophie,id=10,model=ST10000NM0086-2AA101,name=Spin_Retry_Count,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=0i,threshold=97i,value=100i,worst=100i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--CK,host=sophie,id=12,model=ST10000NM0086-2AA101,name=Power_Cycle_Count,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=53i,threshold=20i,value=100i,worst=100i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--CK,host=sophie,id=184,model=ST10000NM0086-2AA101,name=End-to-End_Error,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=0i,threshold=99i,value=100i,worst=100i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--CK,host=sophie,id=187,model=ST10000NM0086-2AA101,name=Reported_Uncorrect,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=0i,threshold=0i,value=100i,worst=100i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--CK,host=sophie,id=188,model=ST10000NM0086-2AA101,name=Command_Timeout,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=0i,threshold=0i,value=100i,worst=100i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O-RCK,host=sophie,id=189,model=ST10000NM0086-2AA101,name=High_Fly_Writes,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=68i,threshold=0i,value=32i,worst=32i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=Past,flags=-O---K,host=sophie,id=190,model=ST10000NM0086-2AA101,name=Airflow_Temperature_Cel,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=35i,threshold=40i,value=65i,worst=39i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--CK,host=sophie,id=191,model=ST10000NM0086-2AA101,name=G-Sense_Error_Rate,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=11417i,threshold=0i,value=95i,worst=95i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--CK,host=sophie,id=192,model=ST10000NM0086-2AA101,name=Power-Off_Retract_Count,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=555i,threshold=0i,value=100i,worst=100i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--CK,host=sophie,id=193,model=ST10000NM0086-2AA101,name=Load_Cycle_Count,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=605i,threshold=0i,value=100i,worst=100i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O---K,host=sophie,id=194,model=ST10000NM0086-2AA101,name=Temperature_Celsius,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=35i,threshold=0i,value=35i,worst=61i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O-RC-,host=sophie,id=195,model=ST10000NM0086-2AA101,name=Hardware_ECC_Recovered,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=344184i,threshold=0i,value=100i,worst=64i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--C-,host=sophie,id=197,model=ST10000NM0086-2AA101,name=Current_Pending_Sector,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=0i,threshold=0i,value=100i,worst=100i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=----C-,host=sophie,id=198,model=ST10000NM0086-2AA101,name=Offline_Uncorrectable,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=0i,threshold=0i,value=100i,worst=100i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-OSRCK,host=sophie,id=199,model=ST10000NM0086-2AA101,name=UDMA_CRC_Error_Count,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=0i,threshold=0i,value=200i,worst=200i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=PO---K,host=sophie,id=200,model=ST10000NM0086-2AA101,name=Multi_Zone_Error_Rate,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=0i,threshold=1i,value=100i,worst=100i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=------,host=sophie,id=240,model=ST10000NM0086-2AA101,name=Head_Flying_Hours,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=48344677i,threshold=0i,value=100i,worst=253i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=------,host=sophie,id=241,model=ST10000NM0086-2AA101,name=Total_LBAs_Written,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=140759859579i,threshold=0i,value=100i,worst=253i 1610529464000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=------,host=sophie,id=242,model=ST10000NM0086-2AA101,name=Total_LBAs_Read,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,raw_value=399844847220i,threshold=0i,value=100i,worst=253i 1610529464000000000
> smart_device,capacity=10000831348736,device=sg3,enabled=Enabled,host=sophie,model=ST10000NM0086-2AA101,serial_no=ZA29PM9W,wwn=5000c500b375b078 exit_status=32i,health_ok=true,read_error_rate=344184i,seek_error_rate=427333388i,temp_c=35i,udma_crc_errors=0i 1610529464000000000
```
As do any of the examples previously documented above.  
However, the following doesn't work:

`devices = [ "/dev/sg3 -d areca,1/2", "/dev/sg3 -d areca,7/2" ]`



### Actual behavior:

The following is the output from specifying more than one drive within the areca array:

```
root@sophie:/etc/telegraf# telegraf  --test| grep smart
2021-01-13T09:20:36Z I! Starting Telegraf 1.17.0
2021-01-13T09:20:36Z I! Using config file: /etc/telegraf/telegraf.conf
> smart_device,device=sg3,host=sophie exit_status=2i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=PO-R--,host=sophie,id=1,model=HGST\ HDN721010ALE604,name=Raw_Read_Error_Rate,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=0i,threshold=16i,value=100i,worst=100i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=--S---,host=sophie,id=2,model=HGST\ HDN721010ALE604,name=Throughput_Performance,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=92i,threshold=54i,value=135i,worst=135i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=POS---,host=sophie,id=3,model=HGST\ HDN721010ALE604,name=Spin_Up_Time,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=406i,threshold=24i,value=175i,worst=175i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--C-,host=sophie,id=4,model=HGST\ HDN721010ALE604,name=Start_Stop_Count,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=99i,threshold=0i,value=100i,worst=100i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=PO--CK,host=sophie,id=5,model=HGST\ HDN721010ALE604,name=Reallocated_Sector_Ct,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=0i,threshold=5i,value=100i,worst=100i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O-R--,host=sophie,id=7,model=HGST\ HDN721010ALE604,name=Seek_Error_Rate,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=0i,threshold=67i,value=100i,worst=100i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=--S---,host=sophie,id=8,model=HGST\ HDN721010ALE604,name=Seek_Time_Performance,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=18i,threshold=20i,value=128i,worst=128i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--C-,host=sophie,id=9,model=HGST\ HDN721010ALE604,name=Power_On_Hours,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=16232i,threshold=0i,value=98i,worst=98i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--C-,host=sophie,id=10,model=HGST\ HDN721010ALE604,name=Spin_Retry_Count,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=0i,threshold=60i,value=100i,worst=100i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--CK,host=sophie,id=12,model=HGST\ HDN721010ALE604,name=Power_Cycle_Count,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=99i,threshold=0i,value=100i,worst=100i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=PO---K,host=sophie,id=22,model=HGST\ HDN721010ALE604,name=Unknown_Attribute,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=100i,threshold=25i,value=100i,worst=100i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--CK,host=sophie,id=192,model=HGST\ HDN721010ALE604,name=Power-Off_Retract_Count,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=823i,threshold=0i,value=100i,worst=100i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--C-,host=sophie,id=193,model=HGST\ HDN721010ALE604,name=Load_Cycle_Count,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=823i,threshold=0i,value=100i,worst=100i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O----,host=sophie,id=194,model=HGST\ HDN721010ALE604,name=Temperature_Celsius,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=36i,threshold=0i,value=166i,worst=166i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O--CK,host=sophie,id=196,model=HGST\ HDN721010ALE604,name=Reallocated_Event_Count,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=0i,threshold=0i,value=100i,worst=100i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O---K,host=sophie,id=197,model=HGST\ HDN721010ALE604,name=Current_Pending_Sector,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=0i,threshold=0i,value=100i,worst=100i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=---R--,host=sophie,id=198,model=HGST\ HDN721010ALE604,name=Offline_Uncorrectable,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=0i,threshold=0i,value=100i,worst=100i 1610529637000000000
> smart_attribute,capacity=10000831348736,device=sg3,enabled=Enabled,fail=-,flags=-O-R--,host=sophie,id=199,model=HGST\ HDN721010ALE604,name=UDMA_CRC_Error_Count,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,raw_value=0i,threshold=0i,value=200i,worst=200i 1610529637000000000
> smart_device,capacity=10000831348736,device=sg3,enabled=Enabled,host=sophie,model=HGST\ HDN721010ALE604,serial_no=1SJS3J5Z,wwn=5000cca26be6b0c3 exit_status=0i,health_ok=true,read_error_rate=0i,seek_error_rate=0i,temp_c=36i,udma_crc_errors=0i 1610529637000000000
```

Only the first drive gets reported on.  Any subsequent drives within that same array don't, with only the 'exit_status=2i" message returned.

Strangely, the following works:
`devices = [ "/dev/sdf", "/dev/sdf", "/dev/sdf", "/dev/sdf" ]`
so it doesn't appear to be a problem with handling duplicate drives.

### Additional info:

Issue #4720 was logged by @sachaz Sept 19,2018 which was identical, except he was using an HP smart array instead of an Areca, but the principle is the same - you must use the '-d' parameter to specify individual drives within the array.

Thanks for this.  I'm struggling to follow your instructions though.  I think you're assuming that I normally run telegraf etc. from my own compiled source code.  I don't.  I simply downloaded the ubuntu telegraf and influx etc. from the website and installed them as per the website instructions.
I've now found the git source and did a 'git clone' into /opt/telegraf and made the change the line in smart.go.  However, I can't run the 'make' command.  I think it's assuming that I have some sort of development environment installed but I don't know what:

```
oot@sophie:/opt/telegraf# make
make: go: Command not found
make: go: Command not found
make: go: Command not found
make[1]: go: Command not found
make[1]: go: Command not found
make[1]: go: Command not found
go mod download
make[1]: go: Command not found
make[1]: *** [Makefile:86: deps] Error 127
make: *** [Makefile:66: all] Error 2
root@sophie:/opt/telegraf# 
```
I'm also guessing that I would need to uninstall the official telegraf ('apt remove telegraf') before replacing it with this new version?

--- a/plugins/inputs/smart/README.md
+++ b/plugins/inputs/smart/README.md
@@ -112,6 +112,14 @@ smartctl --scan -d nvme
   
     ## Timeout for the cli command to complete.
     # timeout = "30s"
+    
+    ## Optionally call smartctl and nvme-cli with a specific concurrency policy.
+    ## By default, smartctl and nvme-cli are called in separate threads (goroutines) to gather disk attributes.
+    ## Some devices (e.g. disks in RAID arrays) may have access limitations that require sequential reading of
+    ## SMART data - one individual array drive at the time. In such case please set this configuration option
+    ## to "sequential" to get readings for all drives.
+    ## valid options: concurrent, sequential
+    # read_method = "concurrent"
 ```
 
 ## Permissions
@@ -235,11 +243,25 @@ the DEVICE (name of the device could be taken from the previous command):
 smartctl --info --health --attributes --tolerance=verypermissive --nocheck NOCHECK --format=brief -d DEVICE
 ```
 
-If you try to gather vendor specific metrics, please provide this commad
+If you try to gather vendor specific metrics, please provide this command
 and replace vendor and device to match your case:
 
 ```sh
 nvme VENDOR smart-log-add DEVICE
+```
+
+If you have specified devices array in configuration file, and Telegraf only shows data from one device, you should
+change the plugin configuration to sequentially gather disk attributes instead of collecting it in separate threads
+(goroutines). To do this find in plugin configuration read_method and change it to sequential:
+
+```toml
+    ## Optionally call smartctl and nvme-cli with a specific concurrency policy.
+    ## By default, smartctl and nvme-cli are called in separate threads (goroutines) to gather disk attributes.
+    ## Some devices (e.g. disks in RAID arrays) may have access limitations that require sequential reading of
+    ## SMART data - one individual array drive at the time. In such case please set this configuration option
+    ## to "sequential" to get readings for all drives.
+    ## valid options: concurrent, sequential
+    read_method = "sequential"
 ```
 
 ## Example SMART Plugin Outputs

--- a/plugins/inputs/smart/smart.go
+++ b/plugins/inputs/smart/smart.go
@@ -43,8 +43,8 @@ var (
 	// PASSED, FAILED, UNKNOWN
 	smartOverallHealth = regexp.MustCompile(`^(SMART overall-health self-assessment test result|SMART Health Status):\s+(\w+).*$`)
 
-	// sasNvmeAttr is a SAS or NVME SMART attribute
-	sasNvmeAttr = regexp.MustCompile(`^([^:]+):\s+(.+)$`)
+	// sasNVMeAttr is a SAS or NVMe SMART attribute
+	sasNVMeAttr = regexp.MustCompile(`^([^:]+):\s+(.+)$`)
 
 	// ID# ATTRIBUTE_NAME          FLAGS    VALUE WORST THRESH FAIL RAW_VALUE
 	//   1 Raw_Read_Error_Rate     -O-RC-   200   200   000    -    0
@@ -53,13 +53,25 @@ var (
 	attribute = regexp.MustCompile(`^\s*([0-9]+)\s(\S+)\s+([-P][-O][-S][-R][-C][-K])\s+([0-9]+)\s+([0-9]+)\s+([0-9-]+)\s+([-\w]+)\s+([\w\+\.]+).*$`)
 
 	//  Additional Smart Log for NVME device:nvme0 namespace-id:ffffffff
+	// nvme version 1.14+ metrics:
+	// ID             KEY                                 Normalized     Raw
+	// 0xab    program_fail_count                             100         0
+
+	// nvme deprecated metric format:
 	//	key                               normalized raw
 	//	program_fail_count              : 100%       0
-	intelExpressionPattern = regexp.MustCompile(`^([\w\s]+):([\w\s]+)%(.+)`)
+
+	// REGEX patter supports deprecated metrics (nvme-cli version below 1.14) and metrics from nvme-cli 1.14 (and above).
+	intelExpressionPattern = regexp.MustCompile(`^([A-Za-z0-9_\s]+)[:|\s]+(\d+)[%|\s]+(.+)`)
 
 	//	vid     : 0x8086
 	//	sn      : CFGT53260XSP8011P
 	nvmeIDCtrlExpressionPattern = regexp.MustCompile(`^([\w\s]+):([\s\w]+)`)
+
+	// Format from nvme-cli 1.14 (and above) gives ID and KEY, this regex is for separating id from key.
+	//  ID			  KEY
+	// 0xab    program_fail_count
+	nvmeIDSeparatePattern = regexp.MustCompile(`^([A-Za-z0-9_]+)(.+)`)
 
 	deviceFieldIds = map[string]string{
 		"1":   "read_error_rate",
@@ -70,7 +82,7 @@ var (
 	}
 
 	// to obtain metrics from smartctl
-	sasNvmeAttributes = map[string]struct {
+	sasNVMeAttributes = map[string]struct {
 		ID    string
 		Name  string
 		Parse func(fields, deviceFields map[string]interface{}, str string) error
@@ -213,9 +225,48 @@ var (
 			Parse: parseTemperatureSensor,
 		},
 	}
-
-	// to obtain Intel specific metrics from nvme-cli
+	// To obtain Intel specific metrics from nvme-cli version 1.14 and above.
 	intelAttributes = map[string]struct {
+		ID    string
+		Name  string
+		Parse func(acc telegraf.Accumulator, fields map[string]interface{}, tags map[string]string, str string) error
+	}{
+		"program_fail_count": {
+			Name: "Program_Fail_Count",
+		},
+		"erase_fail_count": {
+			Name: "Erase_Fail_Count",
+		},
+		"wear_leveling_count": { // previously: "wear_leveling"
+			Name: "Wear_Leveling_Count",
+		},
+		"e2e_error_detect_count": { // previously: "end_to_end_error_detection_count"
+			Name: "End_To_End_Error_Detection_Count",
+		},
+		"crc_error_count": {
+			Name: "Crc_Error_Count",
+		},
+		"media_wear_percentage": { // previously: "timed_workload_media_wear"
+			Name: "Media_Wear_Percentage",
+		},
+		"host_reads": {
+			Name: "Host_Reads",
+		},
+		"timed_work_load": { // previously: "timed_workload_timer"
+			Name: "Timed_Workload_Timer",
+		},
+		"thermal_throttle_status": {
+			Name: "Thermal_Throttle_Status",
+		},
+		"retry_buff_overflow_count": { // previously: "retry_buffer_overflow_count"
+			Name: "Retry_Buffer_Overflow_Count",
+		},
+		"pll_lock_loss_counter": { // previously: "pll_lock_loss_count"
+			Name: "Pll_Lock_Loss_Count",
+		},
+	}
+	// to obtain Intel specific metrics from nvme-cli
+	intelAttributesDeprecatedFormat = map[string]struct {
 		ID    string
 		Name  string
 		Parse func(acc telegraf.Accumulator, fields map[string]interface{}, tags map[string]string, str string) error
@@ -269,6 +320,8 @@ var (
 			Parse: parseBytesWritten,
 		},
 	}
+
+	knownReadMethods = []string{"concurrent", "sequential"}
 )
 
 // Smart plugin reads metrics from storage devices supporting S.M.A.R.T.
@@ -283,6 +336,7 @@ type Smart struct {
 	Devices          []string        `toml:"devices"`
 	UseSudo          bool            `toml:"use_sudo"`
 	Timeout          config.Duration `toml:"timeout"`
+	ReadMethod       string          `toml:"read_method"`
 	Log              telegraf.Logger `toml:"-"`
 }
 
@@ -333,11 +387,20 @@ var sampleConfig = `
 
   ## Timeout for the cli command to complete.
   # timeout = "30s"
+
+  ## Optionally call smartctl and nvme-cli with a specific concurrency policy.
+  ## By default, smartctl and nvme-cli are called in separate threads (goroutines) to gather disk attributes.
+  ## Some devices (e.g. disks in RAID arrays) may have access limitations that require sequential reading of
+  ## SMART data - one individual array drive at the time. In such case please set this configuration option
+  ## to "sequential" to get readings for all drives.
+  ## valid options: concurrent, sequential
+  # read_method = "concurrent"
 `
 
 func newSmart() *Smart {
 	return &Smart{
-		Timeout: config.Duration(time.Second * 30),
+		Timeout:    config.Duration(time.Second * 30),
+		ReadMethod: "concurrent",
 	}
 }
 
@@ -366,6 +429,10 @@ func (m *Smart) Init() error {
 	//if `path_nvme` is not provided in config, try to find nvme binary in PATH
 	if len(m.PathNVMe) == 0 {
 		m.PathNVMe, _ = exec.LookPath("nvme")
+	}
+
+	if !contains(knownReadMethods, m.ReadMethod) {
+		return fmt.Errorf("provided read method `%s` is not valid", m.ReadMethod)
 	}
 
 	err := validatePath(m.PathSmartctl)
@@ -404,9 +471,9 @@ func (m *Smart) Gather(acc telegraf.Accumulator) error {
 			if err != nil {
 				return err
 			}
-			NVMeDevices := distinguishNVMeDevices(devicesFromConfig, scannedNVMeDevices)
+			nvmeDevices := distinguishNVMeDevices(devicesFromConfig, scannedNVMeDevices)
 
-			m.getVendorNVMeAttributes(acc, NVMeDevices)
+			m.getVendorNVMeAttributes(acc, nvmeDevices)
 		}
 		return nil
 	}
@@ -434,28 +501,28 @@ func (m *Smart) scanAllDevices(ignoreExcludes bool) ([]string, []string, error) 
 	}
 
 	// this will return only NVMe devices
-	NVMeDevices, err := m.scanDevices(ignoreExcludes, "--scan", "--device=nvme")
+	nvmeDevices, err := m.scanDevices(ignoreExcludes, "--scan", "--device=nvme")
 	if err != nil {
 		return nil, nil, err
 	}
 
 	// to handle all versions of smartctl this will return only non NVMe devices
-	nonNVMeDevices := difference(devices, NVMeDevices)
-	return NVMeDevices, nonNVMeDevices, nil
+	nonNVMeDevices := difference(devices, nvmeDevices)
+	return nvmeDevices, nonNVMeDevices, nil
 }
 
 func distinguishNVMeDevices(userDevices []string, availableNVMeDevices []string) []string {
-	var NVMeDevices []string
+	var nvmeDevices []string
 
 	for _, userDevice := range userDevices {
-		for _, NVMeDevice := range availableNVMeDevices {
+		for _, availableNVMeDevice := range availableNVMeDevices {
 			// double check. E.g. in case when nvme0 is equal nvme0n1, will check if "nvme0" part is present.
-			if strings.Contains(NVMeDevice, userDevice) || strings.Contains(userDevice, NVMeDevice) {
-				NVMeDevices = append(NVMeDevices, userDevice)
+			if strings.Contains(availableNVMeDevice, userDevice) || strings.Contains(userDevice, availableNVMeDevice) {
+				nvmeDevices = append(nvmeDevices, userDevice)
 			}
 		}
 	}
-	return NVMeDevices
+	return nvmeDevices
 }
 
 // Scan for S.M.A.R.T. devices from smartctl
@@ -506,69 +573,86 @@ func excludedDev(excludes []string, deviceLine string) bool {
 func (m *Smart) getAttributes(acc telegraf.Accumulator, devices []string) {
 	var wg sync.WaitGroup
 	wg.Add(len(devices))
-
 	for _, device := range devices {
-		go gatherDisk(acc, m.Timeout, m.UseSudo, m.Attributes, m.PathSmartctl, m.Nocheck, device, &wg)
+		switch m.ReadMethod {
+		case "concurrent":
+			go m.gatherDisk(acc, device, &wg)
+		case "sequential":
+			m.gatherDisk(acc, device, &wg)
+		default:
+			wg.Done()
+		}
 	}
 
 	wg.Wait()
 }
 
 func (m *Smart) getVendorNVMeAttributes(acc telegraf.Accumulator, devices []string) {
-	NVMeDevices := getDeviceInfoForNVMeDisks(acc, devices, m.PathNVMe, m.Timeout, m.UseSudo)
+	nvmeDevices := getDeviceInfoForNVMeDisks(acc, devices, m.PathNVMe, m.Timeout, m.UseSudo)
 
 	var wg sync.WaitGroup
 
-	for _, device := range NVMeDevices {
+	for _, device := range nvmeDevices {
 		if contains(m.EnableExtensions, "auto-on") {
+			// nolint:revive // one case switch on purpose to demonstrate potential extensions
 			switch device.vendorID {
 			case intelVID:
 				wg.Add(1)
-				go gatherIntelNVMeDisk(acc, m.Timeout, m.UseSudo, m.PathNVMe, device, &wg)
+				switch m.ReadMethod {
+				case "concurrent":
+					go gatherIntelNVMeDisk(acc, m.Timeout, m.UseSudo, m.PathNVMe, device, &wg)
+				case "sequential":
+					gatherIntelNVMeDisk(acc, m.Timeout, m.UseSudo, m.PathNVMe, device, &wg)
+				default:
+					wg.Done()
+				}
 			}
 		} else if contains(m.EnableExtensions, "Intel") && device.vendorID == intelVID {
 			wg.Add(1)
-			go gatherIntelNVMeDisk(acc, m.Timeout, m.UseSudo, m.PathNVMe, device, &wg)
+			switch m.ReadMethod {
+			case "concurrent":
+				go gatherIntelNVMeDisk(acc, m.Timeout, m.UseSudo, m.PathNVMe, device, &wg)
+			case "sequential":
+				gatherIntelNVMeDisk(acc, m.Timeout, m.UseSudo, m.PathNVMe, device, &wg)
+			default:
+				wg.Done()
+			}
 		}
 	}
 	wg.Wait()
 }
 
 func getDeviceInfoForNVMeDisks(acc telegraf.Accumulator, devices []string, nvme string, timeout config.Duration, useSudo bool) []nvmeDevice {
-	var NVMeDevices []nvmeDevice
+	var nvmeDevices []nvmeDevice
 
 	for _, device := range devices {
-		vid, sn, mn, err := gatherNVMeDeviceInfo(nvme, device, timeout, useSudo)
+		newDevice, err := gatherNVMeDeviceInfo(nvme, device, timeout, useSudo)
 		if err != nil {
 			acc.AddError(fmt.Errorf("cannot find device info for %s device", device))
 			continue
 		}
-		newDevice := nvmeDevice{
-			name:         device,
-			vendorID:     vid,
-			model:        mn,
-			serialNumber: sn,
-		}
-		NVMeDevices = append(NVMeDevices, newDevice)
+		nvmeDevices = append(nvmeDevices, newDevice)
 	}
-	return NVMeDevices
+	return nvmeDevices
 }
 
-func gatherNVMeDeviceInfo(nvme, device string, timeout config.Duration, useSudo bool) (string, string, string, error) {
+func gatherNVMeDeviceInfo(nvme, deviceName string, timeout config.Duration, useSudo bool) (device nvmeDevice, err error) {
 	args := []string{"id-ctrl"}
-	args = append(args, strings.Split(device, " ")...)
+	args = append(args, strings.Split(deviceName, " ")...)
 	out, err := runCmd(timeout, useSudo, nvme, args...)
 	if err != nil {
-		return "", "", "", err
+		return device, err
 	}
 	outStr := string(out)
-
-	vid, sn, mn, err := findNVMeDeviceInfo(outStr)
-
-	return vid, sn, mn, err
+	device, err = findNVMeDeviceInfo(outStr)
+	if err != nil {
+		return device, err
+	}
+	device.name = deviceName
+	return device, nil
 }
 
-func findNVMeDeviceInfo(output string) (string, string, string, error) {
+func findNVMeDeviceInfo(output string) (nvmeDevice, error) {
 	scanner := bufio.NewScanner(strings.NewReader(output))
 	var vid, sn, mn string
 
@@ -580,7 +664,7 @@ func findNVMeDeviceInfo(output string) (string, string, string, error) {
 			matches[2] = strings.TrimSpace(matches[2])
 			if matches[1] == "vid" {
 				if _, err := fmt.Sscanf(matches[2], "%s", &vid); err != nil {
-					return "", "", "", err
+					return nvmeDevice{}, err
 				}
 			}
 			if matches[1] == "sn" {
@@ -591,7 +675,13 @@ func findNVMeDeviceInfo(output string) (string, string, string, error) {
 			}
 		}
 	}
-	return vid, sn, mn, nil
+
+	newDevice := nvmeDevice{
+		vendorID:     vid,
+		model:        mn,
+		serialNumber: sn,
+	}
+	return newDevice, nil
 }
 
 func gatherIntelNVMeDisk(acc telegraf.Accumulator, timeout config.Duration, usesudo bool, nvme string, device nvmeDevice, wg *sync.WaitGroup) {
@@ -619,10 +709,31 @@ func gatherIntelNVMeDisk(acc telegraf.Accumulator, timeout config.Duration, uses
 		tags["model"] = device.model
 		tags["serial_no"] = device.serialNumber
 
-		if matches := intelExpressionPattern.FindStringSubmatch(line); len(matches) > 3 {
-			matches[1] = strings.TrimSpace(matches[1])
+		// Create struct to initialize later with intel attributes.
+		var (
+			attr = struct {
+				ID    string
+				Name  string
+				Parse func(acc telegraf.Accumulator, fields map[string]interface{}, tags map[string]string, str string) error
+			}{}
+			attrExists bool
+		)
+
+		if matches := intelExpressionPattern.FindStringSubmatch(line); len(matches) > 3 && len(matches[1]) > 1 {
+			// Check if nvme shows metrics in deprecated format or in format with ID.
+			// Based on that, an attribute map with metrics is chosen.
+			// If string has more than one character it means it has KEY there, otherwise it's empty string ("").
+			if separatedIDAndKey := nvmeIDSeparatePattern.FindStringSubmatch(matches[1]); len(strings.TrimSpace(separatedIDAndKey[2])) > 1 {
+				matches[1] = strings.TrimSpace(separatedIDAndKey[2])
+				attr, attrExists = intelAttributes[matches[1]]
+			} else {
+				matches[1] = strings.TrimSpace(matches[1])
+				attr, attrExists = intelAttributesDeprecatedFormat[matches[1]]
+			}
+
 			matches[3] = strings.TrimSpace(matches[3])
-			if attr, ok := intelAttributes[matches[1]]; ok {
+
+			if attrExists {
 				tags["name"] = attr.Name
 				if attr.ID != "" {
 					tags["id"] = attr.ID
@@ -641,18 +752,18 @@ func gatherIntelNVMeDisk(acc telegraf.Accumulator, timeout config.Duration, uses
 	}
 }
 
-func gatherDisk(acc telegraf.Accumulator, timeout config.Duration, usesudo, collectAttributes bool, smartctl, nocheck, device string, wg *sync.WaitGroup) {
+func (m *Smart) gatherDisk(acc telegraf.Accumulator, device string, wg *sync.WaitGroup) {
 	defer wg.Done()
 	// smartctl 5.41 & 5.42 have are broken regarding handling of --nocheck/-n
-	args := []string{"--info", "--health", "--attributes", "--tolerance=verypermissive", "-n", nocheck, "--format=brief"}
+	args := []string{"--info", "--health", "--attributes", "--tolerance=verypermissive", "-n", m.Nocheck, "--format=brief"}
 	args = append(args, strings.Split(device, " ")...)
-	out, e := runCmd(timeout, usesudo, smartctl, args...)
+	out, e := runCmd(m.Timeout, m.UseSudo, m.PathSmartctl, args...)
 	outStr := string(out)
 
 	// Ignore all exit statuses except if it is a command line parse error
 	exitStatus, er := exitStatus(e)
 	if er != nil {
-		acc.AddError(fmt.Errorf("failed to run command '%s %s': %s - %s", smartctl, strings.Join(args, " "), e, outStr))
+		acc.AddError(fmt.Errorf("failed to run command '%s %s': %s - %s", m.PathSmartctl, strings.Join(args, " "), e, outStr))
 		return
 	}
 
@@ -712,7 +823,7 @@ func gatherDisk(acc telegraf.Accumulator, timeout config.Duration, usesudo, coll
 		tags := map[string]string{}
 		fields := make(map[string]interface{})
 
-		if collectAttributes {
+		if m.Attributes {
 			//add power mode
 			keys := [...]string{"device", "model", "serial_no", "wwn", "capacity", "enabled", "power"}
 			for _, key := range keys {
@@ -724,8 +835,8 @@ func gatherDisk(acc telegraf.Accumulator, timeout config.Duration, usesudo, coll
 
 		attr := attribute.FindStringSubmatch(line)
 		if len(attr) > 1 {
-			// attribute has been found, add it only if collectAttributes is true
-			if collectAttributes {
+			// attribute has been found, add it only if m.Attributes is true
+			if m.Attributes {
 				tags["id"] = attr[1]
 				tags["name"] = attr[2]
 				tags["flags"] = attr[3]
@@ -758,8 +869,8 @@ func gatherDisk(acc telegraf.Accumulator, timeout config.Duration, usesudo, coll
 			}
 		} else {
 			// what was found is not a vendor attribute
-			if matches := sasNvmeAttr.FindStringSubmatch(line); len(matches) > 2 {
-				if attr, ok := sasNvmeAttributes[matches[1]]; ok {
+			if matches := sasNVMeAttr.FindStringSubmatch(line); len(matches) > 2 {
+				if attr, ok := sasNVMeAttributes[matches[1]]; ok {
 					tags["name"] = attr.Name
 					if attr.ID != "" {
 						tags["id"] = attr.ID
@@ -774,8 +885,8 @@ func gatherDisk(acc telegraf.Accumulator, timeout config.Duration, usesudo, coll
 						continue
 					}
 					// if the field is classified as an attribute, only add it
-					// if collectAttributes is true
-					if collectAttributes {
+					// if m.Attributes is true
+					if m.Attributes {
 						acc.AddFields("smart_attribute", fields, tags)
 					}
 				}
@@ -972,13 +1083,13 @@ func parseTemperatureSensor(fields, _ map[string]interface{}, str string) error 
 	return nil
 }
 
-func validatePath(path string) error {
-	pathInfo, err := os.Stat(path)
+func validatePath(filePath string) error {
+	pathInfo, err := os.Stat(filePath)
 	if os.IsNotExist(err) {
-		return fmt.Errorf("provided path does not exist: [%s]", path)
+		return fmt.Errorf("provided path does not exist: [%s]", filePath)
 	}
 	if mode := pathInfo.Mode(); !mode.IsRegular() {
-		return fmt.Errorf("provided path does not point to a regular file: [%s]", path)
+		return fmt.Errorf("provided path does not point to a regular file: [%s]", filePath)
 	}
 	return nil
 }

--- a/plugins/inputs/smart/smart_test.go
+++ b/plugins/inputs/smart/smart_test.go
@@ -24,11 +24,11 @@ func TestGatherAttributes(t *testing.T) {
 			if args[0] == "--info" && args[7] == "/dev/ada0" {
 				return []byte(mockInfoAttributeData), nil
 			} else if args[0] == "--info" && args[7] == "/dev/nvme0" {
-				return []byte(smartctlNvmeInfoData), nil
+				return []byte(smartctlNVMeInfoData), nil
 			} else if args[0] == "--scan" && len(args) == 1 {
 				return []byte(mockScanData), nil
 			} else if args[0] == "--scan" && len(args) >= 2 && args[1] == "--device=nvme" {
-				return []byte(mockScanNvmeData), nil
+				return []byte(mockScanNVMeData), nil
 			}
 		}
 		return nil, errors.New("command not found")
@@ -45,7 +45,7 @@ func TestGatherAttributes(t *testing.T) {
 		s.PathSmartctl = "smartctl"
 		s.PathNVMe = ""
 
-		t.Run("Only non nvme device", func(t *testing.T) {
+		t.Run("Only non NVMe device", func(t *testing.T) {
 			s.Devices = []string{"/dev/ada0"}
 			var acc testutil.Accumulator
 
@@ -62,7 +62,7 @@ func TestGatherAttributes(t *testing.T) {
 				acc.AssertContainsTaggedFields(t, "smart_device", test.fields, test.tags)
 			}
 		})
-		t.Run("Only nvme device", func(t *testing.T) {
+		t.Run("Only NVMe device", func(t *testing.T) {
 			s.Devices = []string{"/dev/nvme0"}
 			var acc testutil.Accumulator
 
@@ -71,9 +71,75 @@ func TestGatherAttributes(t *testing.T) {
 			require.NoError(t, err)
 			assert.Equal(t, 32, acc.NFields(), "Wrong number of fields gathered")
 
-			testutil.RequireMetricsEqual(t, testSmartctlNvmeAttributes, acc.GetTelegrafMetrics(),
+			testutil.RequireMetricsEqual(t, testSmartctlNVMeAttributes, acc.GetTelegrafMetrics(),
 				testutil.SortMetrics(), testutil.IgnoreTime())
 		})
+	})
+}
+
+func TestGatherInParallelMode(t *testing.T) {
+	s := newSmart()
+	s.Attributes = true
+	s.PathSmartctl = "smartctl"
+	s.PathNVMe = "nvmeIdentifyController"
+	s.EnableExtensions = append(s.EnableExtensions, "auto-on")
+	s.Devices = []string{"/dev/nvme0"}
+
+	runCmd = func(timeout config.Duration, sudo bool, command string, args ...string) ([]byte, error) {
+		if len(args) > 0 {
+			if args[0] == "--info" && args[7] == "/dev/ada0" {
+				return []byte(mockInfoAttributeData), nil
+			} else if args[0] == "--info" && args[7] == "/dev/nvmeIdentifyController" {
+				return []byte(smartctlNVMeInfoData), nil
+			} else if args[0] == "--scan" && len(args) == 1 {
+				return []byte(mockScanData), nil
+			} else if args[0] == "--scan" && len(args) >= 2 && args[1] == "--device=nvme" {
+				return []byte(mockScanNVMeData), nil
+			} else if args[0] == "intel" && args[1] == "smart-log-add" {
+				return []byte(nvmeIntelInfoDataMetricsFormat), nil
+			} else if args[0] == "id-ctrl" {
+				return []byte(nvmeIdentifyController), nil
+			}
+		}
+		return nil, errors.New("command not found")
+	}
+
+	t.Run("Gather NVMe device info in goroutine", func(t *testing.T) {
+		acc := &testutil.Accumulator{}
+		s.ReadMethod = "concurrent"
+
+		err := s.Gather(acc)
+		require.NoError(t, err)
+
+		result := acc.GetTelegrafMetrics()
+		testutil.RequireMetricsEqual(t, testIntelNVMeNewFormatAttributes, result,
+			testutil.SortMetrics(), testutil.IgnoreTime())
+	})
+
+	t.Run("Gather NVMe device info sequentially", func(t *testing.T) {
+		acc := &testutil.Accumulator{}
+		s.ReadMethod = "sequential"
+
+		err := s.Gather(acc)
+		require.NoError(t, err)
+
+		result := acc.GetTelegrafMetrics()
+		testutil.RequireMetricsEqual(t, testIntelNVMeNewFormatAttributes, result,
+			testutil.SortMetrics(), testutil.IgnoreTime())
+	})
+
+	t.Run("Gather NVMe device info - not known read method", func(t *testing.T) {
+		acc := &testutil.Accumulator{}
+		s.ReadMethod = "horizontally"
+
+		err := s.Init()
+		require.Error(t, err)
+
+		err = s.Gather(acc)
+		require.NoError(t, err)
+
+		result := acc.GetTelegrafMetrics()
+		testutil.RequireMetricsEqual(t, []telegraf.Metric{}, result)
 	})
 }
 
@@ -90,9 +156,9 @@ func TestGatherNoAttributes(t *testing.T) {
 			} else if args[0] == "--info" && args[7] == "/dev/ada0" {
 				return []byte(mockInfoAttributeData), nil
 			} else if args[0] == "--info" && args[7] == "/dev/nvme0" {
-				return []byte(smartctlNvmeInfoData), nil
+				return []byte(smartctlNVMeInfoData), nil
 			} else if args[0] == "--scan" && args[1] == "--device=nvme" {
-				return []byte(mockScanNvmeData), nil
+				return []byte(mockScanNVMeData), nil
 			}
 		}
 		return nil, errors.New("command not found")
@@ -111,7 +177,7 @@ func TestGatherNoAttributes(t *testing.T) {
 		for _, test := range testsAda0Device {
 			acc.AssertContainsTaggedFields(t, "smart_device", test.fields, test.tags)
 		}
-		for _, test := range testNvmeDevice {
+		for _, test := range testNVMeDevice {
 			acc.AssertContainsTaggedFields(t, "smart_device", test.fields, test.tags)
 		}
 	})
@@ -122,6 +188,16 @@ func TestExcludedDev(t *testing.T) {
 	assert.Equal(t, false, excludedDev([]string{}, "/dev/pass6 -d atacam"), "Shouldn't be excluded.")
 	assert.Equal(t, false, excludedDev([]string{"/dev/pass6"}, "/dev/pass1 -d atacam"), "Shouldn't be excluded.")
 }
+
+var (
+	sampleSmart = Smart{
+		PathSmartctl: "",
+		Nocheck:      "",
+		Attributes:   true,
+		UseSudo:      true,
+		Timeout:      config.Duration(time.Second * 30),
+	}
+)
 
 func TestGatherSATAInfo(t *testing.T) {
 	runCmd = func(timeout config.Duration, sudo bool, command string, args ...string) ([]byte, error) {
@@ -134,7 +210,8 @@ func TestGatherSATAInfo(t *testing.T) {
 	)
 
 	wg.Add(1)
-	gatherDisk(acc, config.Duration(time.Second*30), true, true, "", "", "", wg)
+
+	sampleSmart.gatherDisk(acc, "", wg)
 	assert.Equal(t, 101, acc.NFields(), "Wrong number of fields gathered")
 	assert.Equal(t, uint64(20), acc.NMetrics(), "Wrong number of metrics gathered")
 }
@@ -150,7 +227,7 @@ func TestGatherSATAInfo65(t *testing.T) {
 	)
 
 	wg.Add(1)
-	gatherDisk(acc, config.Duration(time.Second*30), true, true, "", "", "", wg)
+	sampleSmart.gatherDisk(acc, "", wg)
 	assert.Equal(t, 91, acc.NFields(), "Wrong number of fields gathered")
 	assert.Equal(t, uint64(18), acc.NMetrics(), "Wrong number of metrics gathered")
 }
@@ -166,7 +243,7 @@ func TestGatherHgstSAS(t *testing.T) {
 	)
 
 	wg.Add(1)
-	gatherDisk(acc, config.Duration(time.Second*30), true, true, "", "", "", wg)
+	sampleSmart.gatherDisk(acc, "", wg)
 	assert.Equal(t, 6, acc.NFields(), "Wrong number of fields gathered")
 	assert.Equal(t, uint64(4), acc.NMetrics(), "Wrong number of metrics gathered")
 }
@@ -182,7 +259,7 @@ func TestGatherHtSAS(t *testing.T) {
 	)
 
 	wg.Add(1)
-	gatherDisk(acc, config.Duration(time.Second*30), true, true, "", "", "", wg)
+	sampleSmart.gatherDisk(acc, "", wg)
 
 	testutil.RequireMetricsEqual(t, testHtsasAtributtes, acc.GetTelegrafMetrics(), testutil.SortMetrics(), testutil.IgnoreTime())
 }
@@ -198,7 +275,7 @@ func TestGatherSSD(t *testing.T) {
 	)
 
 	wg.Add(1)
-	gatherDisk(acc, config.Duration(time.Second*30), true, true, "", "", "", wg)
+	sampleSmart.gatherDisk(acc, "", wg)
 	assert.Equal(t, 105, acc.NFields(), "Wrong number of fields gathered")
 	assert.Equal(t, uint64(26), acc.NMetrics(), "Wrong number of metrics gathered")
 }
@@ -214,14 +291,14 @@ func TestGatherSSDRaid(t *testing.T) {
 	)
 
 	wg.Add(1)
-	gatherDisk(acc, config.Duration(time.Second*30), true, true, "", "", "", wg)
+	sampleSmart.gatherDisk(acc, "", wg)
 	assert.Equal(t, 74, acc.NFields(), "Wrong number of fields gathered")
 	assert.Equal(t, uint64(15), acc.NMetrics(), "Wrong number of metrics gathered")
 }
 
-func TestGatherNvme(t *testing.T) {
+func TestGatherNVMe(t *testing.T) {
 	runCmd = func(timeout config.Duration, sudo bool, command string, args ...string) ([]byte, error) {
-		return []byte(smartctlNvmeInfoData), nil
+		return []byte(smartctlNVMeInfoData), nil
 	}
 
 	var (
@@ -230,15 +307,15 @@ func TestGatherNvme(t *testing.T) {
 	)
 
 	wg.Add(1)
-	gatherDisk(acc, config.Duration(time.Second*30), true, true, "", "", "nvme0", wg)
+	sampleSmart.gatherDisk(acc, "nvme0", wg)
 
-	testutil.RequireMetricsEqual(t, testSmartctlNvmeAttributes, acc.GetTelegrafMetrics(),
+	testutil.RequireMetricsEqual(t, testSmartctlNVMeAttributes, acc.GetTelegrafMetrics(),
 		testutil.SortMetrics(), testutil.IgnoreTime())
 }
 
-func TestGatherIntelNvme(t *testing.T) {
+func TestGatherIntelNVMeMetrics(t *testing.T) {
 	runCmd = func(timeout config.Duration, sudo bool, command string, args ...string) ([]byte, error) {
-		return []byte(nvmeIntelInfoData), nil
+		return []byte(nvmeIntelInfoDataMetricsFormat), nil
 	}
 
 	var (
@@ -255,17 +332,40 @@ func TestGatherIntelNvme(t *testing.T) {
 	gatherIntelNVMeDisk(acc, config.Duration(time.Second*30), true, "", device, wg)
 
 	result := acc.GetTelegrafMetrics()
-	testutil.RequireMetricsEqual(t, testIntelInvmeAttributes, result,
+	testutil.RequireMetricsEqual(t, testIntelNVMeNewFormatAttributes, result,
+		testutil.SortMetrics(), testutil.IgnoreTime())
+}
+
+func TestGatherIntelNVMeDeprecatedFormatMetrics(t *testing.T) {
+	runCmd = func(timeout config.Duration, sudo bool, command string, args ...string) ([]byte, error) {
+		return []byte(nvmeIntelInfoDataDeprecatedMetricsFormat), nil
+	}
+
+	var (
+		acc    = &testutil.Accumulator{}
+		wg     = &sync.WaitGroup{}
+		device = nvmeDevice{
+			name:         "nvme0",
+			model:        mockModel,
+			serialNumber: mockSerial,
+		}
+	)
+
+	wg.Add(1)
+	gatherIntelNVMeDisk(acc, config.Duration(time.Second*30), true, "", device, wg)
+
+	result := acc.GetTelegrafMetrics()
+	testutil.RequireMetricsEqual(t, testIntelNVMeAttributes, result,
 		testutil.SortMetrics(), testutil.IgnoreTime())
 }
 
 func Test_findVIDFromNVMeOutput(t *testing.T) {
-	vid, sn, mn, err := findNVMeDeviceInfo(nvmeIdentifyController)
+	device, err := findNVMeDeviceInfo(nvmeIdentifyController)
 
 	assert.Nil(t, err)
-	assert.Equal(t, "0x8086", vid)
-	assert.Equal(t, "CVFT5123456789ABCD", sn)
-	assert.Equal(t, "INTEL SSDPEDABCDEFG", mn)
+	assert.Equal(t, "0x8086", device.vendorID)
+	assert.Equal(t, "CVFT5123456789ABCD", device.serialNumber)
+	assert.Equal(t, "INTEL SSDPEDABCDEFG", device.model)
 }
 
 func Test_checkForNVMeDevices(t *testing.T) {
@@ -293,7 +393,7 @@ func Test_difference(t *testing.T) {
 
 func Test_integerOverflow(t *testing.T) {
 	runCmd = func(timeout config.Duration, sudo bool, command string, args ...string) ([]byte, error) {
-		return []byte(smartctlNvmeInfoDataWithOverflow), nil
+		return []byte(smartctlNVMeInfoDataWithOverflow), nil
 	}
 
 	var (
@@ -303,7 +403,8 @@ func Test_integerOverflow(t *testing.T) {
 
 	t.Run("If data raw_value is out of int64 range, there should be no metrics for that attribute", func(t *testing.T) {
 		wg.Add(1)
-		gatherDisk(acc, config.Duration(time.Second*30), true, true, "", "", "nvme0", wg)
+
+		sampleSmart.gatherDisk(acc, "nvme0", wg)
 
 		result := acc.GetTelegrafMetrics()
 		testutil.RequireMetricsEqual(t, testOverflowAttributes, result,
@@ -656,7 +757,7 @@ var (
 	mockModel  = "INTEL SSDPEDABCDEFG"
 	mockSerial = "CVFT5123456789ABCD"
 
-	testSmartctlNvmeAttributes = []telegraf.Metric{
+	testSmartctlNVMeAttributes = []telegraf.Metric{
 		testutil.MustMetric("smart_device",
 			map[string]string{
 				"device":    "nvme0",
@@ -1045,7 +1146,7 @@ var (
 		},
 	}
 
-	testNvmeDevice = []struct {
+	testNVMeDevice = []struct {
 		fields map[string]interface{}
 		tags   map[string]string
 	}{
@@ -1063,7 +1164,7 @@ var (
 		},
 	}
 
-	testIntelInvmeAttributes = []telegraf.Metric{
+	testIntelNVMeAttributes = []telegraf.Metric{
 		testutil.MustMetric("smart_attribute",
 			map[string]string{
 				"device":    "nvme0",
@@ -1257,11 +1358,146 @@ var (
 			time.Now(),
 		),
 	}
+
+	testIntelNVMeNewFormatAttributes = []telegraf.Metric{
+		testutil.MustMetric("smart_attribute",
+			map[string]string{
+				"device":    "nvme0",
+				"serial_no": mockSerial,
+				"model":     mockModel,
+				"name":      "Program_Fail_Count",
+			},
+			map[string]interface{}{
+				"raw_value": 0,
+			},
+			time.Now(),
+		),
+		testutil.MustMetric("smart_attribute",
+			map[string]string{
+				"device":    "nvme0",
+				"serial_no": mockSerial,
+				"model":     mockModel,
+				"name":      "Erase_Fail_Count",
+			},
+			map[string]interface{}{
+				"raw_value": 0,
+			},
+			time.Now(),
+		),
+		testutil.MustMetric("smart_attribute",
+			map[string]string{
+				"device":    "nvme0",
+				"serial_no": mockSerial,
+				"model":     mockModel,
+				"name":      "Wear_Leveling_Count",
+			},
+			map[string]interface{}{
+				"raw_value": int64(700090417315),
+			},
+			time.Now(),
+		),
+		testutil.MustMetric("smart_attribute",
+			map[string]string{
+				"device":    "nvme0",
+				"serial_no": mockSerial,
+				"model":     mockModel,
+				"name":      "End_To_End_Error_Detection_Count",
+			},
+			map[string]interface{}{
+				"raw_value": 0,
+			},
+			time.Now(),
+		),
+		testutil.MustMetric("smart_attribute",
+			map[string]string{
+				"device":    "nvme0",
+				"serial_no": mockSerial,
+				"model":     mockModel,
+				"name":      "Crc_Error_Count",
+			},
+			map[string]interface{}{
+				"raw_value": 13,
+			},
+			time.Now(),
+		),
+		testutil.MustMetric("smart_attribute",
+			map[string]string{
+				"device":    "nvme0",
+				"serial_no": mockSerial,
+				"model":     mockModel,
+				"name":      "Media_Wear_Percentage",
+			},
+			map[string]interface{}{
+				"raw_value": 552,
+			},
+			time.Now(),
+		),
+		testutil.MustMetric("smart_attribute",
+			map[string]string{
+				"device":    "nvme0",
+				"serial_no": mockSerial,
+				"model":     mockModel,
+				"name":      "Host_Reads",
+			},
+			map[string]interface{}{
+				"raw_value": 73,
+			},
+			time.Now(),
+		),
+		testutil.MustMetric("smart_attribute",
+			map[string]string{
+				"device":    "nvme0",
+				"serial_no": mockSerial,
+				"model":     mockModel,
+				"name":      "Timed_Workload_Timer",
+			},
+			map[string]interface{}{
+				"raw_value": int64(2343038),
+			},
+			time.Now(),
+		),
+		testutil.MustMetric("smart_attribute",
+			map[string]string{
+				"device":    "nvme0",
+				"serial_no": mockSerial,
+				"model":     mockModel,
+				"name":      "Thermal_Throttle_Status",
+			},
+			map[string]interface{}{
+				"raw_value": 0,
+			},
+			time.Now(),
+		),
+		testutil.MustMetric("smart_attribute",
+			map[string]string{
+				"device":    "nvme0",
+				"serial_no": mockSerial,
+				"model":     mockModel,
+				"name":      "Retry_Buffer_Overflow_Count",
+			},
+			map[string]interface{}{
+				"raw_value": 0,
+			},
+			time.Now(),
+		),
+		testutil.MustMetric("smart_attribute",
+			map[string]string{
+				"device":    "nvme0",
+				"serial_no": mockSerial,
+				"model":     mockModel,
+				"name":      "Pll_Lock_Loss_Count",
+			},
+			map[string]interface{}{
+				"raw_value": 0,
+			},
+			time.Now(),
+		),
+	}
 	// smartctl --scan
 	mockScanData = `/dev/ada0 -d atacam # /dev/ada0, ATA device`
 
 	// smartctl --scan -d nvme
-	mockScanNvmeData = `/dev/nvme0 -d nvme # /dev/nvme0, NVMe device`
+	mockScanNVMeData = `/dev/nvme0 -d nvme # /dev/nvme0, NVMe device`
 
 	// smartctl --info --health --attributes --tolerance=verypermissive -n standby --format=brief [DEVICE]
 	mockInfoAttributeData = `smartctl 6.5 2016-05-07 r4318 [Darwin 16.4.0 x86_64] (local build)
@@ -1670,7 +1906,7 @@ Selective self-test flags (0x0):
 	After scanning selected spans, do NOT read-scan remainder of disk.
 If Selective self-test is pending on power-up, resume after 0 minute delay.
 `
-	smartctlNvmeInfoData = `smartctl 6.5 2016-05-07 r4318 [x86_64-linux-4.1.27-gvt-yocto-standard] (local build)
+	smartctlNVMeInfoData = `smartctl 6.5 2016-05-07 r4318 [x86_64-linux-4.1.27-gvt-yocto-standard] (local build)
 Copyright (C) 2002-16, Bruce Allen, Christian Franke, www.smartmontools.org
 
 === START OF INFORMATION SECTION ===
@@ -1720,14 +1956,14 @@ Temperature Sensor 7: 44 C
 Temperature Sensor 8: 43 C
 `
 
-	smartctlNvmeInfoDataWithOverflow = `
+	smartctlNVMeInfoDataWithOverflow = `
 Temperature Sensor 1: 9223372036854775808 C
 Temperature Sensor 2: -9223372036854775809 C
 Temperature Sensor 3: 9223372036854775807 C
 Temperature Sensor 4: -9223372036854775808 C
 `
 
-	nvmeIntelInfoData = `Additional Smart Log for NVME device:nvme0 namespace-id:ffffffff
+	nvmeIntelInfoDataDeprecatedMetricsFormat = `Additional Smart Log for NVME device:nvme0 namespace-id:ffffffff
 key                               normalized raw
 program_fail_count              : 100%       0
 erase_fail_count                : 100%       0
@@ -1742,6 +1978,20 @@ retry_buffer_overflow_count     : 100%       0
 pll_lock_loss_count             : 100%       0
 nand_bytes_written              :   0%       sectors: 0
 host_bytes_written              :   0%       sectors: 0
+`
+	nvmeIntelInfoDataMetricsFormat = `Additional Smart Log for NVME device:nvme0n1 namespace-id:ffffffff
+ID             KEY                                 Normalized     Raw
+0xab    program_fail_count                             100         0
+0xac    erase_fail_count                               100         0
+0xad    wear_leveling_count                            100         700090417315
+0xb8    e2e_error_detect_count                         100         0
+0xc7    crc_error_count                                100         13
+0xe2    media_wear_percentage                          100         552
+0xe3    host_reads                                     100         73
+0xe4    timed_work_load                                100         2343038
+0xea    thermal_throttle_status                        100         0
+0xf0    retry_buff_overflow_count                      100         0
+0xf3    pll_lock_loss_counter                          100         0
 `
 
 	nvmeIdentifyController = `NVME Identify Controller:

